File: typescript/packages/bench/src/repositories/invoiceRepository.ts
/** Read side over issued invoices. */

import type { Repository } from "@atelier/core";
import type { Invoice } from "@atelier/core";

/** Read side over issued invoices. */
export class InvoiceRepository implements Repository<number, Invoice> {
    constructor(private readonly records: readonly Invoice[] = []) {}

    find(id: number): Invoice | undefined {
        return this.records.find((record) => record.id === id);
    }

    all(): readonly Invoice[] {
        return this.records;
    }

    count(): number {
        return this.records.length;
    }

    /** Invoices still carrying a balance. */
    unpaid(): readonly Invoice[] {
        return this.records.filter((invoice) => !invoice.paid);
    }
}

File: rust/crates/atelier-bench/src/repositories/invoice_repository.rs
//! Read side over issued invoices.

use atelier_core::contracts::repository::Repository;
use atelier_core::models::invoice::Invoice;

/// Read side over issued invoices.
#[derive(Debug, Clone, Default)]
pub struct InvoiceRepository {
    records: Vec<Invoice>,
}

impl InvoiceRepository {
    pub fn new(records: Vec<Invoice>) -> Self {
        Self { records }
    }

    /// Invoices still carrying a balance.
    pub fn unpaid(&self) -> Vec<&Invoice> {
        self.records
            .iter()
            .filter(|invoice| !invoice.paid)
            .collect()
    }
}

impl Repository for InvoiceRepository {
    type Id = u32;
    type Record = Invoice;

    fn find(&self, id: Self::Id) -> Option<&Self::Record> {
        self.records.iter().find(|record| record.id == id)
    }

    fn all(&self) -> &[Self::Record] {
        &self.records
    }
}
